### PR TITLE
Core: Introduce before/after hooks for modules

### PR DESCRIPTION
**Pull request by trentmwillis**
As discussed in #893. Included tests for the following in support of this feature:

- [x] `before` runs before `beforeEach`
- [x] `after` runs after `afterEach`
- [x] `after` only runs after last test in module
- [x] `before` only runs before first test in module
- [x] first test supports `before` with a promise
- [x] last test supports `after` with a promise
- [x] first test supports `before` with an assert.async
- [x] last test supports `after` with an assert.async
- [x] support nested module `before` hooks
- [x] support nested module `after` hooks
- [x] context changes made in `before` are applied to all tests

### CODE REVIEW

**Comment by trentmwillis** [comment]
@leobalter I think this is ready for review. Tried my best to keep the implementation within existing constructs, but required a few new properties to get all the features/behavior this should support.

File: src/core.js [hunk 1]
```diff
@@ -56,11 +58,14 @@ extend( QUnit, {
 			var module = {
 				name: moduleName,
 				parentModule: parentModule,
-				tests: []
+				tests: [],
+				testsRun: 0
```
Comment (leobalter) [inline src/core.js]: this might be my only concern on this PR. Exposing a new detail on the API conflicts with the plans for the EventEmitter changes (#882).

Although, the `tests` array [will still prevail](https://github.com/js-reporters/js-reporters/issues/30) and I wonder if we can set an equivalent property on the [reference object for the test on that array](https://github.com/jquery/qunit/pull/919/files#diff-f6c2ded91c9881cdc31b5bfcabb54c3bR25). Doing that, we would have the tradeoff to filter the array, but it would save us a new API exposure.

File: test/main/async.js [hunk 2]
```diff
@@ -233,6 +233,35 @@ QUnit.test( "fails if callback is called more than callback call count", functio
...
+
+		assert.throws(function() {
+			overDone();
+		}, new RegExp( "Too many calls to the `assert.async` callback" ) );
+
+		done();
+	}
+});
+
+QUnit.test( "before", function( /* assert */ ) {
```
Comment (leobalter) [inline test/main/async.js]: we may omit the parameter comment.

**Comment by leobalter** [comment]
The tests look great and I like the implementation.

My concerns remained only on the API as in https://github.com/jquery/qunit/pull/919#discussion_r50030019

Other noticed minor style issues for missing spaces between parenthesis, which are not blockers.

File: src/core.js (hunk 1 above)
Comment (trentmwillis) [inline src/core.js]: Looking at the ["events details types" issue](https://github.com/js-reporters/js-reporters/issues/30), I think we would be able to replace `module.testsRun` with `module.status.total` when it is implemented since it looks like it will be similar.

I would also be fine with setting an equivalent property on the `tests` array, though it feels a little less correct.

Comment (leobalter) [inline src/core.js]: > Looking at the "events details types" issue, I think we would be able to replace module.testsRun with module.status.total when it is implemented since it looks like it will be similar.

:+1: that would be great!

Comment (trentmwillis) [inline src/core.js]: In the interim, should we proceed with this current implementation? (After I clean up the style issues, of course.)

Comment (leobalter) [inline src/core.js]: Yes, sure.

We will need to have the API docs on http://github.com/jquery/api.qunitjs.com to be able to land it, would you mind to help on that part too?

Comment (trentmwillis) [inline src/core.js]: I'll open a PR documenting the new `before`/`after` hooks for `module`. Is there anything else I should document?

**Comment by trentmwillis** [comment]
Updated the PR (think I got all the style issues).

File: src/core.js (hunk 1 above)
Comment (leobalter) [inline src/core.js]: I believe they'll be enough for now, QUnit docs will probably face a whole logging details review on a near future.

File: src/test.js [hunk 3]
```diff
@@ -132,10 +134,22 @@ Test.prototype = {
...
 		return function runHook() {
+			if ( hookName === "before" ) {
+				if ( hookOwner.testsRun !== 0 ) {
+					return;
+				}
+
+				test.preserveEnvironment = true;
+			}
+
+			if ( hookName === "after" && hookOwner.testsRun !== numberOfTests(hookOwner) - 1 ) {
```
Comment (leobalter) [inline src/test.js]: a small style issue here `numberOfTests(hookOwner)` => `numberOfTests( hookOwner )

Comment (trentmwillis) [inline src/test.js]: Ah, missed one.

File: src/test.js [hunk 4]
```diff
@@ -74,8 +74,10 @@ Test.prototype = {
 		config.current = this;
 
 		if ( this.module.testEnvironment ) {
+			delete this.module.testEnvironment.before;
 			delete this.module.testEnvironment.beforeEach;
 			delete this.module.testEnvironment.afterEach;
+			delete this.module.testEnvironment.after;
```
Comment (gibson042) [inline src/test.js]: This may require treating the change as breaking. @leobalter?

**Comment by gibson042** [comment]
I wonder if this should cover edge cases like [delaying the definition of a test](https://jsfiddle.net/obuj0czg/). They're a bit pathological, but it would still be bad to run an `after` hook more than once.

**Comment by trentmwillis** [comment]
@gibson042 is that a real use case? If so, I don't really see a realistic way to handle it other than making an explicit declaration of the number of tests your module should have, since we need advanced knowledge of whether or not another test is expected to run.

File: src/test.js (hunk 4 above)
Comment (leobalter) [inline src/test.js]: I don't think so. We have this delete statement only because we're introducing new properties to the modules' hook: `before` and `after`. Anyone previously extending that module with these keywords would get their tests affected. 

That same could happen when we moved `setup` and `teardown` to `beforeEach` and `afterEach`. Similarly we could mention the introduction of `QUnit.skip` to the API that could  interfere with any other use of custom made methods.

**Comment by leobalter** [comment]
> @trentmwillis: is that a real use case?

That example reflects tests added through async files loading, as in AMD. A dirty check could use a global variable to tell they are added.

Anyway, I also believe it's a developer responsibility to organize well tests introduced asynchronously. An update to the example - where I only added a single line: reveals a structure antipattern: https://jsfiddle.net/obuj0czg/1/

**Comment by trentmwillis** [comment]
Okay, makes sense. Should I attempt to cover that use case? Based on the example you gave, it probably doesn't make much sense without guaranteeing what module a test belongs to

**Comment by leobalter** [comment]
> Should I attempt to cover that use case?

I believe a simple timeout example is enough to cover it up, testing it works properly on async operations and it does not leak anywhere else. AMD cases will be the responsibility for those who implements it.

File: src/test.js (hunk 4 above)
Comment (gibson042) [inline src/test.js]: > Anyone previously extending that module with these keywords would get their tests affected.

Exactly. Code like `QUnit.module( name, { after: new Date(2016, 0, 1) }, … )`, _which is currently supported_, will be broken. This is why I abhor the conflation of module options and test context in a single parameter, but there it is. And it's different from adding `QUnit.skip` because `QUnit` is our object, but that parameter is explicitly wide open for users, minus our exceptions.

Anyway, while maybe not a huge deal, this is technically breaking.

**Comment by gibson042** [comment]
> Anyway, I also believe it's a developer responsibility to organize well tests introduced asynchronously. An update to the example - where I only added a single line: reveals a structure antipattern: https://jsfiddle.net/obuj0czg/1/

Right, pathological. :smile:

**Comment by gibson042** [comment]
Sorry about the noise. But my point was that we should add documentation and assertions that lock down how `after` behaves in such cases (either "invoked exactly once, and therefore ignored delayed tests" or "invoked whenever the module's test queue is empty, and therefore after delayed tests").

**Comment by trentmwillis** [comment]
> "invoked whenever the module's test queue is empty, and therefore after delayed tests"

Sorry if I'm missing something obvious, but going back to my earlier point, if a test definition is delayed, then it may not get added to the module's test queue until after the module has already emptied it's test queue. How do we notify the module that even though it has ran all of it's tests, it should not run the `after` hook until more tests are added and ran?

**Comment by gibson042** [comment]
> How do we notify the module that even though it has ran all of it's tests, it should not run the `after` hook until more tests are added and ran?

We don't. We just document and assert our behavior, which will take one of two forms:
* Invoke `after` exactly once. Tests added after its invocation will have to manage their own cleanup.
* Invoke `after` whenever the module's test queue becomes empty. Tests added after its first invocation will result in successive invocations, and users employing such patterns must structure it to accommodate.

I'm just forcing the above as an explicit choice, rather than an accident of implementation.

**Comment by trentmwillis** [comment]
> I'm just forcing the above as an explicit choice, rather than an accident of implementation.

Okay, cleared that all up. Currently the implementation means your second option is true (it will result in successive invocations), though I think the first option makes more sense. This is because I imagine `after` callbacks will often pair with `before` callbacks, which are only run once.

If that makes sense, I'll add a test for that here and document it in the documentation PR.

**Comment by leobalter** [comment]
> If that makes sense,  I'll add a test for that here and document it in the documentation PR.

@trentmwillis ++. 

do you mind? It would be great to have it tested and documented.

**Comment by trentmwillis** [comment]
@leobalter updated with a test and updated the doc PR as well. Travis failed but it looks like it's for the same reason the master build is failing currently.

**Comment by leobalter** [comment]
I'm sick of browserstack-runner timeouts.

File: src/test.js (hunk 4 above)
Comment (leobalter) [inline src/test.js]: > This is why I abhor the conflation of module options and test context in a single parameter

We should discuss a way to solve this before 2.0. Maybe we need to isolate the hooks from the environment or event remove the possibility to define env values on this object (let `this` do its job).

**Comment by trentmwillis** [comment]
@leobalter @gibson042 is the only remaining question whether or not we should consider this breaking? If so, I'd be fine with feature-flagging this somehow or delaying merge until 2.0.

**Comment by leobalter** [comment]
@trentmwillis, I really want to merge and release a new version with it, but I might need to release a major version.

I'm running on other necessary changes (the warning api) to release the next major, so releasing this feature won't be a problem.

**Comment by trentmwillis** [comment]
@leobalter a major version sounds good. Will let this sit until ready then.

**Comment by stefanpenner** [comment]
@trentmwillis good work!

**Comment by trentmwillis** [comment]
@leobalter this has been rebased and should be good to go. Care to do a once over since it's been awhile?
